fuel-relayer/src/finalization_queue.rs

```diff
@@ -265,17 +266,42 @@ impl FinalizationQueue {
         }
         self.remove_bundled_reverted_events();
 
+        //TODO to be paranoid, recheck every block and all events got from eth client.
+
+        let mut validators: HashMap<Address, Option<Address>> = HashMap::new();
         while let Some(diff) = self.pending.front_mut() {
             if diff.da_height > finalized_da_height {
                 break;
             }
             info!("flush eth log:{:?} diff:{:?}", diff.da_height, diff);
-            //TODO to be paranoid, recheck events got from eth client.
+
+            let validator_diff: HashMap<Address, ValidatorDiff> = diff
+                .validators
+                .iter()
+                .map(|(val, &new_consensus_key)| {
+                    let previous_consensus_key = match validators.entry(*val) {
+                        Entry::Occupied(mut entry) => {
+                            core::mem::replace(entry.get_mut(), new_consensus_key)
+                        }
+                        Entry::Vacant(entry) => {
+                            entry.insert(new_consensus_key);
+                            self.validators.set.get(val).and_then(|(_, i)| *i)
```

Comment:
[nit] isn't this the same thing as `new_consensus_key`?

```suggestion
                            new_consensus_key
```

Reply:
Also, if this was vacant before, shouldn't previous_consensus_key be None?

Reply:
if it is inside `validator` it means that the previous diff changed it in some way, but if it is not inside `validator` we need to fetch original (previous) value and that is why we call `self.validators`

Reply:
ahh ok I was confused by all the different validator sets and what they were being used for. So IIUC, to determine the previous key, we first check a cross-block cache scoped to this function call for any current keys, and if that's empty we fallback to the latest validator set snapshot loaded from the database.

I don't see anything in this function to update the validator set or reload it except during initialization. So if there were several calls to this commit_diffs function, could we end up using the wrong previous keys once the `validators` hashmap drops between calls?

Reply:
Yep, that 's the logic.

update for `self.validator.set` is done in function below, `bump_set_to_da_height` where i read from db and update it, this is not the best but functions were already made before last iteration of relayer was devised